R/dplyr.R

```diff
@@ -25,3 +26,47 @@ mutate.skim_df <- function(.data, ...) {
   mutated <- NextMethod("mutate")
   rebuild_skim_obj(mutated, .data)
 }
+
+#'  Transmute a skim_df
+#'  
+#' [dplyr::transmute()] older than version 0.8.0 drops attributes, but we need to keep 
+#' them around for other skim behaviors and consistency with newer versions of dplyr. 
+#' Otherwise the behavior is exactly the same as dplyr 0.8.0. For more
+#' information, see https://github.com/tidyverse/dplyr/issues/3429. We also need to
+#' retain the two metadata columns (`variable` and `type`) that are the defining 
+#' features of a skim_df object.
+#'
+#' @param .data A `skim_df`, which behaves like a `tbl.`
+#' @param ... Name-value pairs of expressions, each with length 1 or the same
+#'   length as the number of rows in the group (if using [group_by()] or in the
+#'   entire input (if not using groups). The name of each argument will be the
+#'   name of a new variable, and the value will be its corresponding value. 
+#'   Use `NULL` value in `mutate` to drop a variable.  New variables overwrite
+#'   existing variables of the same name.
+#'   @return A `skim_df` object, which also inherits the class(es) of the input
+#'   data. The `variable`` and `type` variables will be included
+#'   even if not named. 
+#'   In many ways, the object behaves like a [tibble::tibble()].
+#' @seealso [dplyr::transmute()] for the function's expected behavior.
+#' @export
+transmute.skim_df <- function(.data, ...) {
+  transmuted <- NextMethod("transmute")
+  transmuted <- transmuted[!colnames(transmuted) %in% c("variable", "type")]
+  base_skim <- .data[c("variable", "type")]
+  transmuted <- cbind(base_skim, transmuted)
+  rebuild_skim_obj(transmuted, .data)
+}
+
+#' Use dplyr verb select on skim_df objects.
+#' Always includes variable and type columns even if user has not selected them.
+#' @seealso [`dplyr::select()`]
+#' @inheritParams dplyr::select
+#' @return  skim_df object
+#' @export
+select.skim_df <-function (.data, ...) {
+   mc <- match.call()
```

Comment:
This behavior should be covered by our package's function "focus".

To be safe, we could still keep this around, only matching the attributes. Like transmute above, you could warn the user that they were destroying the skim object and revert it to df.

Reply:
Well, we're still stuck then.

Basically select is now just an alias for focus, since they are doing essentially the same thing. If we're going this route, we shouldn't have two functions that behave the same way..

Reply:
Correct,  I see three options.  Clearly we will not have two doing the same thing.  One can  allow selects that don't include variable and type,  but always (?) strip off the skim_df class.  One can prevent removal of variable and type and always keep the skim_df class.   I think a third option is have one function with a parameter that either allows you to or  prevents you from removing them.  Probably there are more possibilities but to me those are the main ones.  And then the question is which one do we call select and which do we call focus.

Reply:
If we don't have something called select and a users uses select, then they are by default allowed to not select variable and type (and any grouping variables!) but they also keep the skim_df class even if  it isn't a skim_df.   So I think at minimum we need a select method to deliver the message and strip off the class.  I don't think we want to sometimes return a skim_df class and other times not.